drivers/can/can_native_linux_adapt.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ * Copyright (c) 2022 Martin Jäger <martin@libre.solar>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Routines setting up the host system. Those are placed in separate file
+ * because there is naming conflicts between host and zephyr network stacks.
+ */
+
+/* Linux host include files. */
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <errno.h>
+#include <string.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <net/if.h>
+#include <linux/can.h>
+#include <linux/can/raw.h>
+
+#include "can_native_linux_priv.h"
+
+int can_native_linux_iface_open(const char *if_name)
+{
+	struct sockaddr_can addr;
+	struct ifreq ifr;
+	int fd, ret = -EINVAL;
+
+	fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+	if (fd < 0) {
+		return -errno;
+	}
+
+	(void)memset(&ifr, 0, sizeof(ifr));
+	(void)memset(&addr, 0, sizeof(addr));
+
+	strncpy(ifr.ifr_name, if_name, IFNAMSIZ);
+
+	ret = ioctl(fd, SIOCGIFINDEX, (void *)&ifr);
+	if (ret < 0) {
+		close(fd);
+		return -errno;
+	}
+
+	addr.can_ifindex = ifr.ifr_ifindex;
+	addr.can_family = PF_CAN;
+
+	ret = bind(fd, (struct sockaddr *)&addr, sizeof(addr));
+	if (ret < 0) {
+		close(fd);
+		return -errno;
+	}
+
+	return fd;
+}
+
+int can_native_linux_iface_close(int fd)
+{
+	return close(fd);
+}
+
+int can_native_linux_wait_data(int fd)
```

Comment:
Note that you are not waiting but polling for data. This function name can be missleading.